File: Week 01/id_303/RotateArray.swift
//
//  RotateArray.swift
//  AlgorithmStudy


//189.旋转数组
import Foundation

class Solution {
    //方法1.取余数
    func rotate(_ nums: inout [Int], _ k: Int) {
        let terns = k % nums.count
        if nums.isEmpty || terns == 0 {
            return
        }
        var newNums = nums
        for i in 0..<nums.count {
            newNums[(i + terns) % nums.count] = nums[i]
        }
        nums = newNums
    }
    //方法2.先把前n-k个数字翻转一下，再把后k个数字翻转一下，最后再把整个数组翻转一下
    func rotateA(_ nums: inout [Int], _ k: Int) {
        let terns = k % nums.count
        if nums.isEmpty || terns == 0 {
            return
        }
        let middle = nums.count - terns
        
        reverse(&nums, s: 0, e: middle - 1)
        reverse(&nums, s: middle, e: nums.count - 1)
        reverse(&nums, s: 0, e: nums.count - 1)
    }
    
    func reverse(_ nums: inout [Int], s: Int, e: Int) {
        var s = s
        var e = e
        while s < e {
            let temp = nums[s]
            nums[s] = nums[e]
            nums[e] = temp
            
            s += 1
            e -= 1
        }
    }
}

File: Week 01/id_713/RotateArray.java
package id_713;

/**
 * 189. 旋转数组
 * 给定一个数组，将数组中的元素向右移动 k 个位置，其中 k 是非负数。
 * 示例 1:
 * 输入: [1,2,3,4,5,6,7] 和 k = 3
 * 输出: [5,6,7,1,2,3,4]
 * 示例 2:
 * 输入: [-1,-100,3,99] 和 k = 2
 * 输出: [3,99,-1,-100]
 */
public class RotateArray {

    /*
    思路:
    1. 整体翻转
    2. 轴之前翻转
    3. 轴之后翻转
     */
    public void rotate(int[] nums, int k) {
        k = k % nums.length;

        reverse(nums, 0, nums.length - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, nums.length - 1);
    }

    private void reverse(int[] nums, int i, int j) {
        int tmp = 0;
        while (i < j) {
            tmp = nums[i];
            nums[i] = j;
            nums[j] = tmp;
            i++;
            j--;
        }
    }
}
